[TOPIC:DTS] dts: enhance enum checking

galak opened the pull request:
Based on pr #18164

ulfalizer committed:
dts: dtlib: Remember the format of assignments

Previously, dtlib just stored the raw 'bytes' value for each property,
along with some markers in Property._markers for phandle and path
references.

Extend Property._markers to also remember where different data blocks
start, so that e.g.

    foo = <1 2 3>, "bar", [00 01];

can be reproduced as written.

Use the new information to reproduce properties as written in
Property.__str__(). This gives good test coverage as well, since the
test suite checks literal __str__() output.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
dts: dtlib/edtlib: Add a syntax-based type-checking system

Property type-checking has been pretty rudimentary until now, only
checking things like the length being divisible by 4 for 'type: array',
and strings being null-terminated. In particular, no checking was done
for 'type: uint8-array', letting

  jedec-id = < 0xc8 0x28 0x17 >;

slip through when

  jedec-id = [ 0xc8 0x28 0x17 ];

was intended.

Fix it by adding a syntax-based type checker:

  1. Add Property.type, which gives a high-level type for the property,
     derived from the markers added in the previous commit.

     This includes types like TYPE_EMPTY ('foo;'),
     TYPE_NUM ('foo = < 3 >;'), TYPE_BYTES ('foo = [ 01 02 ];'),
     TYPE_STRINGS ('foo = "bar", "baz"'),
     TYPE_PHANDLE ('foo = < &bar >;'), and TYPE_COMPOUND (everything not
     recognized).

     See the Property.type docstring in dtlib for more info.

  2. Use the high-level type in
     Property.to_num()/to_string()/to_node()/etc. to verify that the
     property was assigned in an expected way for the type.

     If the assignment looks bad, give a helpful error:

       expected property 'nums' on /foo/bar in some.dts to be assigned
       with 'nums = < (number) (number) ... >', not 'nums = "oops";'

Some other related changes are included as well:

  - There's a new Property.to_bytes() function that works like accessing
    Property.bytes, except with an added check for the value being
    assigned like 'foo = [ ... ]'.

    This function solves problems like the jedec-id one.

  - There's a new Property.to_path() function for fetching the
    referenced node for assignments like 'foo = &node;', with type
    checking. (Strings are accepted too, as long as they give the path
    to an existing node.)

    This function is used for /chosen and /aliases.

  - A new 'type: phandle' type can now be given in bindings, for
    properties that are assigned like 'foo = < &node >;'.

  - Property.__str__() now displays phandles and path references as they
    were written (e.g. '< &foo >' instead of '< 0x1 >', if the
    allocated phandle happened to be 1).

  - Property.to_num() and Property.to_nums() no longer take a 'length'
    parameter, because it makes no sense with the type checking.

  - The global dtlib.to_string() and dtlib.to_strings() functions were
    removed, because they're not that useful.

  - More tests were added, along with misc. minor cleanup in various
    places.

  - Probably other stuff I forgot.

The more strict type checking in dtlib indirectly makes some parts of
edtlib more strict as well (wherever Property.to_*() is used).

Fixes: #18131

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

galak committed:
scripts/dts/edtlib.py: Hoist enum checking before early out

Move the enum checking before we early out for '#' and '-map' properties
so they can benefit from it.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts/dts/edtlib.py: error check enum

Make sure 'enum' is a list in the binding.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,14 @@ def _init_prop(self, name, options):
+        # Handle enum checking
+        enum = options.get("enum")
+        if enum:
+            if type(enum) is not list: enum = [enum]
```
I'd prefer to require it to be a list instead, and error out otherwise.

I think dtschema uses `const:` for this. Maybe we could add that instead?

ulfalizer replied:
From the example-schema.yaml in dtschema:

```
  # Property names starting with '#' must be quoted
  '#interrupt-cells':
    # A simple case where the value must always be '2'.
    # The core schema handles that this must be a single integer.
    const: 2
```

ulfalizer replied:
(Just discovered that dt-schema supports `default:` by the way. See https://github.com/zephyrproject-rtos/zephyr/pull/17928.)

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,13 @@ def _init_prop(self, name, options):
             # 'category: optional' property that wasn't there
             return
 
+        # Handle enum checking
```
Bit redundant comment.

galak replied:
done

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,13 @@ def _init_prop(self, name, options):
+        # Handle enum checking
+        enum = options.get("enum")
+        if enum:
+            if val not in enum:
```
Can do `if enum and val not in enum:`.

galak replied:
Fixed.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -562,14 +569,9 @@ def _init_prop(self, name, options):
-                _err("value ({}) for property ({}) is not in enumerated "
-                     "list {} for node {}".format(val, name, enum, self.name))
-
             prop.enum_index = enum.index(val)
```
Can shorten a bit with

```python
prop.enum_index = None if enum is None else enum.index(val)
```

ulfalizer replied:
The `None` is to distinguish from an empty list by the way (though it's kinda pointless to have an empty list for `enum`).

Another option would be to require `enum` to have at least one element (can check that with `if enum`, as empty lists are falsy).

galak replied:
Done.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -553,6 +553,9 @@ def _init_prop(self, name, options):
         if enum:
+            if type(enum) is not list:
+               _err("enum in {} for property ({}) is not "
+                    "a list".format(self.binding_path, name))
```
Not obligatory, but could consider having this in `_check_binding()` as well.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -553,6 +553,9 @@ def _init_prop(self, name, options):
         # Handle enum checking
         enum = options.get("enum")
         if enum:
+            if type(enum) is not list:
```
`if not isinstance(enum, list)` is more common.

galak replied:
done

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,15 @@ def _init_prop(self, name, options):
             return
 
+        enum = options.get("enum")
+        if enum and not isinstance(enum, list):
```
Thinking about it, we should have this in `_check_binding()`. As it is now, the check is skipped if we return above, and the whole binding should always be checked if possible.

galak replied:
Moved into _check_binding()

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1250,6 +1247,11 @@ def _check_binding(binding, binding_path):
+        enum = options.get("enum")
+        if enum and not isinstance(enum, list):
+            _err("enum in {} for property ({}) is not "
+                 "a list".format(binding_path, prop_name))
```
Could do this to be consistent with the other checks:

```python
if "enum" in options and not isinstance(options["enum"], list):
    _err("enum in {} for property '{}' is not a list"
         .format(binding_path, prop_name))
```

galak replied:
Changed.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,11 @@ def _init_prop(self, name, options):
+        enum = options.get("enum")
+        if enum and val not in enum:
+            _err("value ({}) for property ({}) is not in enumerated "
+                 "list {} for node {}".format(val, name, enum, self.name))
```
`self.path` is better than `self.name`, as it gives the full path to the node.

galak replied:
Changed.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -550,6 +550,11 @@ def _init_prop(self, name, options):
 
+        enum = options.get("enum")
+        if enum and val not in enum:
+            _err("value ({}) for property ({}) is not in enumerated "
```
"value '{}' for property '{}' is not..." maybe, to be consistent (and since the name is essential info).

galak replied:
Done.

ulfalizer committed:
dts: edtlib: Fix outdated doc comment for _init_devices()

No longer takes the DT instance as a parameter.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
dts: dtlib: Fix parsing of character literals

This was allowed due to a misunderstanding:

    foo = 'x';

In reality, 'x' works like an integer literal, and is used like this:

    foo = < 'x' >;

Fix character literal parsing to match the C tools.

Also fix backslash escape parsing to match the C tools exactly
(get_escape_char() in util.c): \<char> should be turned into <char> if
<char> isn't recognized as a special escape character, instead of being
left alone. This fixes parsing of e.g. '\'' (a character literal with a
single quote in it).

Piggyback some more tests for weird property/node names.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

galak commented:
Closing this pr, will merge as part of #18222
